### PR TITLE
staged_ledger: no hashing while validating txns

### PR DESCRIPTION
**Pull request by cmr**
This pulls all of the logic for applying transactions and creating undos from `Coda_base.Ledger` and into `Transaction_logic` functor. We only use it in one new place: the `Transaction_validator`.  The `Transaction_validator` has no notion of `undo` and avoids all hashing.

This prevents a substantial amount of hashing when creating a diff.

### CODE REVIEW

**Comment by psteckler** [comment]
One this PR is done, then we can delete all `remove_accounts_exn` (issue #1584)?

**Comment by cmr** [comment]
> One this PR is done, then we can delete all `remove_accounts_exn` (issue #1584)?

Unfortunately not. The last use of `Ledger.undo` is in staged ledger, in `snarked_ledger` (which works by undoing the transactions that are still pending in the parallel scan state).

File: src/lib/coda_base/transaction_validator.ml [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+    s
+    ^ ": somehow we got a location that isn't present in the underlying ledger"
+
+  let get t = function
+    | Ours key -> Hashtbl.find t.overlay key
+    | Theirs loc -> (
+      match Ledger.get t.base loc with
+      | Some a -> (
+        match Hashtbl.find t.overlay (Account.public_key a) with
+        | None -> Some a
```
Comment (bkase) [inline src/lib/coda_base/transaction_validator.ml]: nit: `Option.value ~default` instead of a match here

Comment (cmr) [reply src/lib/coda_base/transaction_validator.ml]: As usual, `Option.value` has the wrong type. This is Rust's `Option::or_else`.

**Pull request merged**